tests/unit/test_endpoint_provider.py

```diff
@@ -544,3 +553,141 @@ def test_deeply_nested(self):
             ]
         }
         self.assertDictEqual(expected_rules_file, rules_file_dict_read)
+
+
+def provider_with_rules_factory(rules: Dict[str, str]):
+    """Class factory to create a provider charm class with the given rules.
+
+    Args:
+        rules: a mapping between filename and contents.
+    """
+
+    class ProviderCharm(CharmBase):
+        metadata_yaml = textwrap.dedent(
+            """
+            provides:
+              metrics-endpoint:
+                interface: prometheus_scrape
+            """
+        )
+
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args)
+
+            self.sandbox = TempFolderSandbox()
+            alert_rules_path = os.path.join(self.sandbox.root, "alert_rules")
+            for filepath, contents in rules.items():
+                self.sandbox.put_file(os.path.join(alert_rules_path, filepath), contents)
+
+            self.provider = MetricsEndpointProvider(
+                self, jobs=JOBS, alert_rules_path=alert_rules_path
+            )
+
+    return ProviderCharm
+
+
+class TestAlertRulesDoNotContainUnitNumber(unittest.TestCase):
```

Comment:
Why do we need a separate test class here ?

Reply:
As I understand we need two test cases for `MetricsEndpointProvider`

1. A test case to check that `juju_unit` is not added to alert rules and expressions by default. As mentioned above adding two lines to an existing test case meets this requirement.
2. A test case to check that if `juju_unit` is present in an alert rule label and expression then it is not overwritten. 

For 1 you should not need any new test class. For 2 you will need a separate test class, since you need a different set of alert rules who's labels and expressions do include Juju unit topology annotations.

Case 1 corresponds to the name `TestAlertRulesDoNotContainUnitNumber`.

Reply:
I am confused. Feel free to push a commit!